### ISSUE TITLE
RCP should store a member's join date (and maybe more)

### ISSUE THREAD

**Issue by tnorthcutt** [restrictcontentpro/restrict-content-pro#186]
Use case: a site where some members (let's say "basic") should only get access to content added to the site _since their join date_, but other members (let's say "premium") should get access to all content.

Storing a join date on member creation would make this possible (still requiring custom code, of course, but the data would be available).

Complications: what if a user's status goes to cancelled or expired, and then active? In my use-case above, they should have a new "join date", but I can easily see a scenario where the "join date" should be fixed in stone as the date of their original joining. In that case, also having an "active date" could be useful, which would be the date their status changed to active.

I'd love to get feedback from others on this, especially if you've dealt with this or a similar scenario on a production site.

**Comment by pippinsplugins** [comment]
After thinking about it more, I think the date should represent the signup date for their current subscription.

We can access the actual creation date for the account from wp_users->user_registered

**Comment by tnorthcutt** [comment]
@pippinsplugins What would that date be in the following scenarios?

1. Member joins on January 1, 2015.
2. Member expires on March 1, 2015.
3. Member becomes active on June 1, 2015.

1. Member joins on January 1, 2015.
2. Member cancels on March 1, 2015.
3. Member becomes active (re-joins?) on June 1, 2015.

**Comment by bearded-avenger** [comment]
+1 i just came across this need in the new CGC build. I need to grab the number of days a user has been active on this membership, and can only be gotten if we had a member join date. http://cl.ly/image/1i3225300e3v

**Comment by tnorthcutt** [comment]
MemberMouse has a somewhat useful set of options for this. Of course I'm not suggesting we implement an identical solution, but could be inspiration for a good option:
![image](https://cloud.githubusercontent.com/assets/796639/8236688/d6642ea6-159e-11e5-9538-5353f83eec90.png)

**Comment by pippinsplugins** [comment]
@tnorthcutt I think the join date is the date their subscription starts for the first time. That does bring up the question of whether or not the `renewed last` date should also be stroed.

**Comment by carter2422** [comment]
I think it may make sense to store a **Initial Join Date** and **Renewed Date**. Often time I'd like to know when a user first subscribed, and sometimes when they most recently subscribed. 

In my mind the **Initial Join Date** should be set once and only once, much like an account created date. The reason being is it's valuable to know when someone *first* became a paying customer.

**Comment by mindctrl** [comment]
If we want to backfill dates, depends on https://github.com/restrictcontentpro/restrict-content-pro/issues/406.

**Comment by mindctrl** [comment]
Might want to do this after https://github.com/restrictcontentpro/restrict-content-pro/issues/103 too.

**Comment by pippinsplugins** [comment]
:+1:

**Comment by pippinsplugins** [comment]
In order to backfill this, I vote we grab the date of the earliest payment.

We should also store the join date per subscription level.

**Comment by pippinsplugins** [comment]
PR ready for testing.

I've introduced four new methods in the `RCP_Member` class:
- get_joined_date()
- set_joined_date()
- get_renewed_date()
- set_renewed_date()

Dates are set on a per-subscription level basis.

For old accounts created before this change, the joined date is determined by the date of the member's first payment for their current subscription level.

**Comment by pippinsplugins** [comment]
Additional items to consider:
- We could show the joined / renewed date on the member details screen.
- For members that have expired and then come back to renew manually, should we reset the joined date or leave as is?

**Comment by carter2422** [comment]
> For old accounts created before this change, the joined date is determined by the date of the member's first payment for their current subscription level.

This won't work for *a lot* of members. For example, any member that's ever upgraded to a different plan or come back after an absence and joined a different plan. 

Any reason not to just use first payment date, regardless of plan? 

> For members that have expired and then come back to renew manually, should we reset the joined date or leave as is?

I'd vote for the `Join Date` representing the first signup. This is the number that is really valuable, since it lets you determine how long a customer has been around, even if intermittently.

**Comment by pippinsplugins** [comment]
@carter2422 I'm setting the date per subscription specifically to account for cases where access to content is limited by time since registration. Take a scenario will content is made accessible 3 months after registration to a paid plan. If the site gives a free plan with limited access, using a single registration date won't work since a member could sign up for the free plan, wait three months, then upgrade to the paid plan and immediately gain access.

**Comment by carter2422** [comment]
@pippinsplugins might be worth storing the original join date and then a `days on plan` then.

**Comment by pippinsplugins** [comment]
@carter2422 Is the join date in your scenario really any different than the `user_registered` date?

**Comment by carter2422** [comment]
@pippinsplugins yes because `user_registered` date is when a free account is created not necessarily when a paid membership began.

**Comment by pippinsplugins** [comment]
@carter2422 See, that's why we need it stored per level ;)

**Comment by mindctrl** [comment]
I'm a little confused. To summarize the dates we have and/or want:

**User Registered Date:** WordPress account creation date, not always related to a member's subscription date. We do nothing here.

**Initial Join Date:** The *very first time* the member registered for *any* subscription (first payment). Useful for finding out how long a member has been a subscriber (obviously excluding any lapses). It seems like we do not have this here?

**Current Subscription Start Date:** the start date of the current subscription.

**Renewal Date:** the most recent date the current subscription was renewed.

Right? Wrong?

**Comment by pippinsplugins** [comment]
@mindctrl that's right. I'm unsure, however, if we really need to separate user registered date and initial join date. In 99% of cases or more, those will be identical.

**Comment by mindctrl** [comment]
Join dates look good here now. Properly saved during new registrations and upgrades.

**Comment by tnorthcutt** [comment]
_Revisiting this much later..._

FWIW, I'm running into a situation where knowing an initial join date (when someone first became a paying member) would be important, but where instead using the "join date" of their current subscription would be noise.

Many membership sites _want_ members to upgrade to higher tiers/more expensive plans over time. However, if doing so essentially resets a member's join date, that can throw off sophisticated efforts at engagement tracking over time, spotting trends in membership lengths, etc. For those purposes, as soon as someone upgrades their join date becomes meaningless (and misleading).

Would there be any support for reconsidering how this is handled (meaning adding methods to calculate the initial join date) , or is it considered too late at this point?

**Comment by tnorthcutt** [comment]
Also, it just occurred to me that in light of RCP supporting multiple subscriptions in the future, this could get even more confusing/noisy. Maybe that's a good argument in favor of an initial join date 😉

**Comment by mindctrl** [comment]
I think we should be storing an initial join date that never changes.

WP stores a `user_registered` date on the user object, which would work for a lot of sites that only use RCP to create accounts, but that's not foolproof for sites that allow user account creation via other means (EDD, Woo, form plugins, etc) because they may have created an account but weren't members until later. It's a good guess though, if all other means fail to provide the data.

Other means of trying to determine initial join date would be:
- date of member's first payment (wouldn't work for free members since we only started recording $0 payments in 2.9)
- custom query to get all a user's meta where `meta_key LIKE rcp_joined_date_%` and then sort then based on the date in the `meta_value` so you can get the first join date regardless of current subscription.
- parsing member notes for the join date. member notes are annoying as hell to parse though. I went down that road last week and the inconsistencies in formatting make it difficult to do it confidently. Some new lines are formatted one way, some another, and we don't enforce a format for user-entered notes so the possibilities are endless.
- any ways I'm not thinking about?

**Comment by tnorthcutt** [comment]
@mindctrl good thoughts.

IMO parsing notes is a non-starter; as you mentioned it's fragile and not dependable due to user-entered notes.

Custom query and then sorting by date sounds promising.

This may be overly complicated, but: would it make any sense to distinguish between _paid_ joined date and _overall_ joined date? That may be overthinking things/trying to plan for too many possible scenarios.

Re: custom query and then parsing, we could also introduce an intermediate method (`RCP_Member->get_joined_dates()`) which returns an array with subscription IDs for keys and join dates for values (or vice versa); then e.g. `RCP_Member->get_initial_join_date()` could consume that and return the oldest one. I could see that intermediate method being useful to extension developers.

**Comment by nosegraze** [comment]
I definitely agree with adding an initial join date. Just a note that the custom query route won't be perfect. For a while we were overwriting the join date when users manually renewed the same subscription level. https://github.com/restrictcontentpro/restrict-content-pro/issues/1286 Though that might still be the best in a bunch of not-great options.

**Comment by chrisvanpatten** [comment]
It almost seems like what would be most useful is some kind of consistent audit log that can be accessed via API to pull in exactly when certain events occur. Meta doesn't really feel like the best place for that information over the long term, especially on larger installs or when you need more dates/events tracked. I know user notes are somewhat used for that, but as everyone noted you can't really access it programatically.

Somewhat tempted to build this as a 3rd party plugin… 😬

**Comment by nosegraze** [comment]
@chrisvanpatten This may interest you: https://github.com/restrictcontentpro/restrict-content-pro/issues/1104

**Comment by tnorthcutt** [comment]
@mindctrl FYI, I did some testing and the custom query method for `rcp_joined_date_%` usermeta method may not be reliable.

I have a local test site with three subscription levels, with IDs of `1` , `2`, and `3`.

When changing from one subscription level to another, the corresponding usermeta that gets updated is the one for the subscription level that a member is _leaving_.

I suspect this is a bug in `RCP_Member->set_joined_date()`. That method uses `RCP_Member->get_subscription_id()` to get the ID for use in `rcp_joined_date_%` usermeta. If the joined date is being set before the member's new subscription ID is set, then this behavior makes sense – so I think that must be the case.

Obviously for new signups, the correct joined date usermeta is getting set. This leads me to conclude that for upgrades & downgrades, the joined date is getting set before the new subscription ID is set, while for new signups the subscription ID is set first.

I _think_ this points to a bug in `registration-functions.php`.

I'm going to go ahead and copy most of this text into a new issue to specifically document this bug so it doesn't get missed in this feature request issue 😄
